pkg/certificate/cert_manager/provider.go

```
@@ -0,0 +1,251 @@
+		privateKey, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse private key: %w", err)
+		}
+	}
+
+	return privateKey, nil
+}
+
+// checkKeyPair checks if the private key matches the certificate.
+func checkKeyPair(cert *x509.Certificate, privateKey crypto.PrivateKey) error {
+	switch key := privateKey.(type) {
+	case *rsa.PrivateKey:
+		// Check if the private key matches the certificate by validating the public key
+		pub := cert.PublicKey.(*rsa.PublicKey)
+		if !key.PublicKey.Equal(pub) {
+			return errors.New("private key does not match the public key in the certificate")
+		}
+	case *ecdsa.PrivateKey:
+		// Check if the private key matches the certificate by validating the public key
+		pub := cert.PublicKey.(*ecdsa.PublicKey)
+		if !key.PublicKey.Equal(pub) {
+			return errors.New("private key does not match the public key in the certificate")
+		}
+	default:
+		return fmt.Errorf("unsupported private key type: %T", key)
+	}
+
+	return nil
+}
+
+func (cm *Provider) EnsureCertificateSecret(
+	ctx context.Context,
+	secretName string,
+	namespace string,
+	cfg *interfaces.Config) error {
+	issuerName, found := cfg.ExtraConfig["issuerName"].(string)
+	if !found || len(strings.TrimSpace(issuerName)) == 0 {
+		return errors.New("issuerName is missing in ExtraConfig")
+	}
+	issuerKind, found := cfg.ExtraConfig["issuerKind"].(string)
+	if !found || len(strings.TrimSpace(issuerKind)) == 0 {
+		return errors.New("issuerKind is missing in ExtraConfig")
+	}
+
+	checkCertSecret, valErr := cm.ValidateCertificateSecret(ctx, secretName, namespace, cfg)
+	if valErr != nil {
+		return fmt.Errorf("invalid certificate secret: %s present in namespace: %s, please delete and try again.\nError: %s",
+			secretName, namespace, valErr)
+	}
+	if checkCertSecret {
+		return fmt.Errorf("valid certificate secret: %s already present in namespace: %s , skipping Certificate creation",
+			secretName, namespace)
```

This should not be an error. If you insist on returning an error, you should return a predefined error something like `ErrCertAlreadyExisted`, so that the caller can double check the returned error.